### PR TITLE
Build haddock in CI

### PR DESCRIPTION
**Pull request by michaelpj**
- Misc improvements around specification of compilers
- Build haddock derivations for components that have them

This should ensure nothing gets published with broken haddock. Let's see if we get away with it...

### CODE REVIEW

**Comment by michaelpj** [comment]
Okay, this works but it's not very beautiful, wdyt @andreabedini ?

**Comment by andreabedini** [comment]
It looks ok to me. I allowed myself to make a little change to get the haddock published by Hydra. Maybe, I might need to pass them to the aggregate; also I never understood why Hydra is not showing the constituents.

**Comment by andreabedini** [comment]
Finally I figured out!

```
15:39:28 hydra-evaluator.1    | aggregate job ‘jobs.test’ has a constituent ‘/nix/store/04w2zdarr340slal8m9c25ccglvciqx5-cardano-node-lib-cardano-node-8.2.1-haddock.drv’ that doesn't correspond to a Hydra build
15:39:28 hydra-evaluator.1    | aggregate job ‘jobs.test’ has a constituent ‘/nix/store/33yg9xhz0qb4028k2306dyz7bs50smq7-cardano-node-lib-cardano-node-8.2.1.drv’ that doesn't correspond to a Hydra build
15:39:28 hydra-evaluator.1    | aggregate job ‘jobs.test’ has a constituent ‘/nix/store/j36fqba1ypli6sncfsma2910rlrvra1z-cardano-node-exe-cardano-node-8.2.1.drv’ that doesn't correspond to a Hydra build
```

The jobs you aggregate still have to be in the set of hydra jobs!

File: flake.nix [hunk 1]
```diff
@@ -49,64 +117,18 @@
...
-          # using intersectAttrs like this is a cheap way to throw away everything with keys not in
-          # smokeTestPackages
-          smokeTestPkgVersions =
-            builtins.intersectAttrs
-              (lib.genAttrs smokeTestPackages (pkg: { }))
-              (chap-meta.chap-package-latest-versions chap-meta.chap-package-meta);
+
+          builder = import ./nix/builder.nix { inherit pkgs CHaP extraConfig; };
+
+          extraConfig = compiler:
```
Comment (michaelpj) [inline flake.nix:123]: Maybe put this next to `compilerBlacklist`? That way all the stuff about "fixing exceptions" is next to each other.

File: nix/builder.nix [hunk 2]
```diff
@@ -60,31 +70,45 @@ let
...
-        pkgs.releaseTools.aggregate
-          {
+      mkJob = project:
+        let
+          # package components
+          components =
+            haskellLib.mkFlakePackages
+              { ${package-name} = project.hsPkgs.${package-name}; };
+
+          # package components documentation
```
Comment (michaelpj) [inline nix/builder.nix:80]: link to https://github.com/input-output-hk/haskell.nix/issues/1932, which presumably if fixed would mean that `mkFlakePackages` would include this?

**Review by michaelpj — COMMENTED** [review]
Looks fine, although it seems like we had a weird merge at some point.

File: nix/builder.nix [hunk 3]
```diff
@@ -43,11 +42,22 @@ let
...
 
+          documentation: True
+
           extra-packages: ${package-id}
         '';
+
+        modules =
+          extraConfig compiler-nix-name
+          ++ [{
+            postHaddock = ''
```
Comment (michaelpj) [inline nix/builder.nix:53]: Oh, this should presumably be fixed in `haskell.nix` also, can we get an upstream ticket link?

Comment (andreabedini) [reply nix/builder.nix:53]: https://github.com/input-output-hk/haskell.nix/issues/2029

File: nix/builder.nix (hunk 2 above)
Comment (andreabedini) [reply nix/builder.nix:80]: I'll assign that to myself.

**Review by andreabedini — APPROVED** [review]

**Pull request merged**